include/openssl/core_names.h

```diff
@@ -202,7 +208,27 @@ extern "C" {
 
 /* Key Exchange parameters */
 
-#define OSSL_EXCHANGE_PARAM_PAD      "pad" /* uint */
+#define OSSL_EXCHANGE_PARAM_PAD                   "pad" /* uint */
+#define OSSL_EXCHANGE_PARAM_EC_ECDH_COFACTOR_MODE "ecdh-cofactor-mode" /* int */
```

Comment:
Nit: rather than repeat the same string twice, can't you have `OSSL_PKEY_PARAM_ECDH_COFACTOR_MODE` as value?

Reply:
I considered this, and I am not entirely opposed to do it, but my rationale for choosing this repetition (counting on the compiler to reduce the memory cost) is that semantically and programmatically they are different:

- the exchange param can take value -1, 0 or 1
- the key param can take only 0 or 1


It is the job of the documentation to explain this (as a separate PR) but I still fear that an header that depicts one of them as an alias of the other might cause confusion in external developers and is prone to future mistakes on our (or at least mine!) behalf as well.

Reply:
I would make the strings different, then.  It would make sense to keep the name "ecdh-cofactor-mode" for the exchange param (if nothing else because it will be recognised as a name that looks like the legacy control string key "ecdh_cofactor_mode"), but have "ecdh-cofactor-flag" for the key param.

Reply:
I like this proposal!

Are you sure you are fine with "ecdh-cofactor-flag" rather than "ecdh-cofactor-mode-flag"?

I'd prefer the latter as the flag is about enabling/disabling the cofactor_mode and not about having/not_having a cofactor (which without knowledge of the internals would be a plausible cofactor, given that standards about encoding EC keys talk of the cofactor as an optional field in the struct).


Should I also rename the define to match? E.g. `OSSL_PKEY_PARAM_ECDH_COFACTOR_FLAG` (or `*_COFACTOR_MODE_FLAG`).

Reply:
I don't think making if *more* wordy helps matters, especially since "mode" is quite vague as it is.  "use-cofactor-flag", or even "use-cofactor" would be better, because that's exactly the meaning of that flag, is it not?

Reply:
Would a change like this be acceptable to you?

~~~diff
 /* Elliptic Curve Key Parameters */
-#define OSSL_PKEY_PARAM_ECDH_COFACTOR_MODE  "ecdh-cofactor-mode"
+#define OSSL_PKEY_PARAM_USE_COFACTOR_FLAG "use-cofactor-flag"
+#define OSSL_PKEY_PARAM_USE_ECDH_COFACTOR \
+    OSSL_PKEY_PARAM_USE_COFACTOR_FLAG
~~~

I don't want to drop `ecdh_cofactor` from the constant name, as that is a sure way of grepping through the code to find definitions and usage of stuff related to all this madness for people dealing with it, compared to grepping for `cofactor` that would instead give a lot of false positives.

But I like the idea of making `USE_ECDH_COFACTOR` an alias for the more generic `USE_COFACTOR_FLAG`.

Reply:
That looks okay

Reply:
See a2d57bbc68 (for the code changes) and 7fbe952eef (for the doc changes)!

Reply:
Regarding documentation, it seem curious to document the same name string twice, and differently.

Reply:
@levitte does 45eec41fc4 look better?